xarray/tests/test_weighted.py

```diff
@@ -0,0 +1,324 @@
+
+    data = DataArray(data, dims=dims, coords=coords)
+
+    if as_dataset:
+        data = data.to_dataset(name="data")
+
+    if operator == "sum_of_weights":
+        result = getattr(data.weighted(weights), operator)(dim)
+    else:
+        result = getattr(data.weighted(weights), operator)(dim, skipna=skipna)
+
+    expected = expected_weighted(data, weights, dim, skipna, operator)
+
+    assert_allclose(expected, result)
+
+
+@pytest.mark.parametrize("dim", ("dim_0", None))
+@pytest.mark.parametrize("shape_data", ((4,), (4, 4), (4, 4, 4)))
+@pytest.mark.parametrize("shape_weights", ((4,), (4, 4), (4, 4, 4)))
+@pytest.mark.parametrize("operator", ("sum_of_weights", "sum", "mean"))
+@pytest.mark.parametrize("add_nans", (True, False))
+@pytest.mark.parametrize("skipna", (None, True, False))
+@pytest.mark.parametrize("as_dataset", (True, False))
+def test_weighted_operations_different_shapes(
+    dim, shape_data, shape_weights, operator, add_nans, skipna, as_dataset
+):
+
+    weights = DataArray(np.random.randn(*shape_weights))
+
+    data = np.random.randn(*shape_data)
+
+    # add approximately 25 % NaNs
+    if add_nans:
+        c = int(data.size * 0.25)
+        data.ravel()[np.random.choice(data.size, c, replace=False)] = np.NaN
+
+    data = DataArray(data)
+
+    if as_dataset:
+        data = data.to_dataset(name="data")
+
+    if operator == "sum_of_weights":
+        result = getattr(data.weighted(weights), operator)(dim)
+    else:
+        result = getattr(data.weighted(weights), operator)(dim, skipna=skipna)
+
+    expected = expected_weighted(data, weights, dim, skipna, operator)
+
+    assert_allclose(expected, result)
+
+
+@pytest.mark.parametrize("operator", ("sum_of_weights", "sum", "mean"))
+@pytest.mark.parametrize("as_dataset", (True, False))
+@pytest.mark.parametrize("keep_attrs", (True, False, None))
+def test_weighted_operations_keep_attr(operator, as_dataset, keep_attrs):
+
+    weights = DataArray(np.random.randn(2, 2), attrs=dict(attr="weights"))
+    data = DataArray(np.random.randn(2, 2))
+
+    if as_dataset:
+        data = data.to_dataset(name="data")
+
+    data.attrs = dict(attr="weights")
+
+    result = getattr(data.weighted(weights), operator)(keep_attrs=True)
+
+    if operator == "sum_of_weights":
+        assert weights.attrs == result.attrs
+    else:
+        assert data.attrs == result.attrs
+
+    result = getattr(data.weighted(weights), operator)(keep_attrs=None)
+    assert not result.attrs
+
+    result = getattr(data.weighted(weights), operator)(keep_attrs=False)
+    assert not result.attrs
+
+
+@pytest.mark.xfail(reason="xr.Dataset.map does not copy attrs of DataArrays GH: 3595")
+@pytest.mark.parametrize("operator", ("sum", "mean"))
+def test_weighted_operations_keep_attr_da_in_ds(operator):
+    # GH #3595
+
+    weights = DataArray(np.random.randn(2, 2), attrs=dict(attr="weights"))
+    data = DataArray(np.random.randn(4, 4, 4), attrs=dict(attr="data"))
```

Comment:
this should be a `Dataset`? This test should pass now.

Reply:
Yes you are right - I have to do `data = data.to_dataset(name="data")